src/redis-cli.c

```diff
@@ -9820,6 +10021,574 @@ void testHintSuite(char *filename) {
     exit(fail);
 }
 
+/*------------------------------------------------------------------------------
+ * Keystats
+ *--------------------------------------------------------------------------- */
+
+/* key length distribution. */
+
+typedef struct size_dist_entry {
+    unsigned long long size;
+    unsigned long long count;
+} size_dist_entry;
+
+typedef struct size_dist {
+    unsigned long long total_count;
+    unsigned long long total_size;
+    unsigned long long max_size;
+    size_dist_entry *size_dist;
+} size_dist;
+
+/* distribution is an array initialized with last element {UINT64_MAX, 0}
+ * for instance: size_dist_entry distribution[] = { {32, 0}, {256, 0}, {UINT64_MAX, 0} }; */
+static void sizeDistInit(size_dist *dist, size_dist_entry *distribution) {
+    dist->max_size = 0;
+    dist->total_count = 0;
+    dist->total_size = 0;
+    dist->size_dist = distribution;
+}
+
+static void addSizeDist(size_dist *dist, unsigned long long size) {
+    dist->total_count++;
+    dist->total_size += size;
+
+    if (size > dist->max_size) {
+        dist->max_size = size;
+    }
+
+    int j;
+    for (j=0; size > dist->size_dist[j].size; j++);
+    dist->size_dist[j].count++;
+}
+
+static int displayKeyStatsLengthDist(size_dist *dist) {
+    int line_count = 0;
+    unsigned long long total_keys = 0, size;
+    char buf[2][256];
+
+    line_count += cleanPrintfln("Key length     Percentile Total keys");
+    line_count += cleanPrintfln("-------------- ---------- -----------");
+
+    for (int i=0; dist->size_dist[i].size < UINT64_MAX; i++) {
```

Comment:
whether i is out of bounds? i will be greater than the length of size_dist, and theoretically the loop could infinite, size will always be less than UINT64_MAX.
It seems that we still need to put a {0,0} at the end of the `distribution`, `otherwise we won't know its size.

Reply:
The original version was 0 terminated, I think your suggestion to terminate with UINT64_MAX works too. We just have to make sure we enter the last element with UINT64_MAX or 0, depending of what we want to test for.
`size_dist_entry distribution[] = { {32, 0}, {256, 0}, {UINT64_MAX, 0} };`
or
`size_dist_entry distribution[] = { {32, 0}, {256, 0}, {0, 0} };`
When we add a value, we stop when the value is less or equal to the current element, it works because we cannot go above UINT64_MAX.
When we read the count per value, we stop when we reach UINT64_MAX. We read the entire array.
We could write the for loop this way too:
`for (int i=0; dist->size_dist[i].size != UINT64_MAX; i++)`
Maybe we could test with `dist->max_size` to stop earlier, but I am not sure it is going to make the code simpler.
We can pick 0 terminated or UINT64_MAX terminated, I think both can work. We should just pick what makes it easier to read.
Let me know if I missed something that will make the loop to go out of bound.

Reply:
Last commit is back to 0 terminated distribution.